File: packages/unity-package/Runtime/VreenModel.cs
// VreenModel.cs — pure C# data model mirroring io.vreen.core.model (Kotlin).
// Used by Unity / .NET runtimes. No external dependencies.

using System;
using System.Collections.Generic;

namespace Vreen
{
    public static class Versions
    {
        public const string Current = "0.2.1";
        public const string Legacy = "0.1.0";
        public const string World = "0.2.0";
    }

    public enum AssetKind
    {
        Model,
        Texture,
        Hdri,
        Audio,
    }

    [Serializable]
    public class VreenAssetEntry
    {
        public string id;
        public string kind;       // AssetKind string
        public string path;
        public long size;
        public string sha256;
        public string originalName;
        public Dictionary<string, object> meta;

        public AssetKind GetKind() => (AssetKind)Enum.Parse(typeof(AssetKind), kind, true);
    }

    [Serializable]
    public class VreenManifest
    {
        public string version = Versions.Current;
        public string exportedAt;
        public string name;
        public string assetName;
        public string generator;
        public VreenAssetEntry[] assets = Array.Empty<VreenAssetEntry>();
        public string primaryModelId; // null → empty
        public VreenWorldJson world;
    }

    [Serializable]
    public class VreenWorldJson
    {
        public string version = Versions.World;
        public string name;
        public long frame;
        public VreenEntityJson[] entities = Array.Empty<VreenEntityJson>();
    }

    [Serializable]
    public class VreenEntityJson
    {
        public long id;
        public string name;
        public VreenSceneNodeJson sceneNode;
        public VreenComponentJson[] components = Array.Empty<VreenComponentJson>();
    }

    [Serializable]
    public class VreenSceneNodeJson
    {
        public float[] position; // [x,y,z]
        public float[] rotation; // quaternion [x,y,z,w]
        public float[] scale;
    }

    [Serializable]
    public class VreenComponentJson
    {
        public string type;
        public Dictionary<string, object> data;
    }

    [Serializable]
    public class VreenScene
    {
        public string version = Versions.Current;
        public Dictionary<string, object> camera = new();
        public Dictionary<string, object> animation = new() { { "speed", 1.0 } };
        public Dictionary<string, object> environment = new()
        {
            { "preset", "midnight" }, { "exposure", 1.0 },
            { "background", "solid" }, { "backgroundColor", "#000000" },
        };
        public Dictionary<string, object> postFX = new()
        {
            { "bloom", false }, { "bloomIntensity", 0.0 },
            { "chromaticAberration", false }, { "vignette", false }, { "ssao", false },
        };
        public Dictionary<string, Dictionary<string, object>> materials = new();
    }

    public class UnpackedVreen
    {
        public VreenManifest manifest;
        public VreenScene scene;
        public Dictionary<string, byte[]> assets; // id → bytes
        public VreenWorldJson world;
    }

    public class VreenFormatError : Exception
    {
        public VreenFormatError(string message) : base(message) { }
        public VreenFormatError(string message, Exception inner) : base(message, inner) { }
    }
}

File: packages/vreen-core/src/main/kotlin/io/vreen/core/model/VreenModel.kt
package io.vreen.core

import com.fasterxml.jackson.annotation.JsonIgnoreProperties

/** Format version constants. Mirrors src/lib/vreenManifest.ts. */
object Versions {
    const val CURRENT = "0.2.1"
    const val LEGACY = "0.1.0"
    const val WORLD = "0.2.0"
}

/** Asset categories. */
enum class AssetKind(val path: String) {
    MODEL("assets"),
    TEXTURE("assets/textures"),
    HDRI("assets/hdri"),
    AUDIO("assets/audio");

    companion object {
        fun fromString(s: String): AssetKind = entries.first { it.name.equals(s, true) }
    }
}

/** Single asset descriptor (manifest.assets[]). */
data class VreenAssetEntry(
    val id: String,
    val kind: AssetKind,
    val path: String,
    val size: Long,
    val sha256: String? = null,
    val originalName: String? = null,
    val meta: Map<String, Any?>? = null,
) {
    init {
        require(id.isNotBlank()) { "asset id must not be blank" }
        require(!path.startsWith("/")) { "asset path must not start with /" }
        require(!path.contains("..")) { "asset path must not contain '..'" }
        require(!path.contains('\\')) { "asset path must use forward slashes only" }
        require(size >= 0) { "size must be >= 0" }
        if (sha256 != null) {
            require(sha256.length == 64) { "sha256 must be 64 hex chars" }
            require(sha256.all { it.isDigit() || it in 'a'..'f' }) { "sha256 must be lowercase hex" }
        }
    }
}

/** Top-level manifest. */
@JsonIgnoreProperties(ignoreUnknown = true)
data class VreenManifest(
    val version: String = Versions.CURRENT,
    val exportedAt: String,
    val name: String,
    val assetName: String,
    val generator: String,
    val assets: List<VreenAssetEntry>,
    val primaryModelId: String?,
    val world: VreenWorldJson? = null,
) {
    init {
        require(version == Versions.CURRENT) { "manifest.version must be ${Versions.CURRENT}, got $version" }
        require(name.isNotBlank()) { "name must not be blank" }
        if (primaryModelId != null) {
            val m = assets.firstOrNull { it.id == primaryModelId }
            requireNotNull(m) { "primaryModelId $primaryModelId not in assets" }
            require(m.kind == AssetKind.MODEL) { "primaryModelId must reference a model asset" }
        }
    }
}

/** ECS world (data-only, see §6 of vreen-format-spec.md). */
data class VreenWorldJson(
    val version: String = Versions.WORLD,
    val name: String,
    val frame: Long,
    val entities: List<VreenEntityJson>,
)

data class VreenEntityJson(
    val id: Long,
    val name: String,
    val sceneNode: VreenSceneNodeJson,
    val components: List<VreenComponentJson>,
)

data class VreenSceneNodeJson(
    val position: List<Float>,
    val rotation: List<Float>, // quaternion [x, y, z, w]
    val scale: List<Float>,
)

data class VreenComponentJson(
    val type: String,
    val data: Map<String, Any?>,
)

/** Scene sub-record (scene.json). */
data class VreenScene(
    val version: String = Versions.CURRENT,
    val camera: Map<String, Any?> = emptyMap(),
    val animation: Map<String, Any?> = mapOf("speed" to 1.0),
    val environment: Map<String, Any?> = mapOf(
        "preset" to "midnight",
        "exposure" to 1.0,
        "background" to "solid",
        "backgroundColor" to "#000000",
    ),
    val postFX: Map<String, Any?> = mapOf(
        "bloom" to false, "bloomIntensity" to 0.0,
        "chromaticAberration" to false, "vignette" to false, "ssao" to false,
    ),
    val materials: Map<String, Map<String, Any?>> = emptyMap(),
) {
    init {
        require(version == Versions.CURRENT) { "scene.version must be ${Versions.CURRENT}" }
    }
}

/** Unpacked package result. */
data class UnpackedVreen(
    val manifest: VreenManifest,
    val scene: VreenScene,
    val assets: Map<String, ByteArray>, // id → bytes
    val world: VreenWorldJson?,
)
